Clean-up error handler & fix message for unknown API errors

paescuj opened the pull request:
## Scope

- Fix broken logic to get error message of unknown errors (e.g. from a Flow) for admin users
  - Brackets were missing, which resulted in the error object itself being used as the message
  - **Example Flow**
     <table><tr><td><img width="400" src="https://github.com/directus/directus/assets/5363448/d027c70d-ce59-4c3e-aafa-87c722af0f6b"></tr></td></table>
    <b>Before</b>
     <table><tr><td><img width="400" src="https://github.com/directus/directus/assets/5363448/0366cf50-8d23-48c9-a47e-e39ae049d5d5"></tr></td></table>
    <b>After</b>
    <table><tr><td><img width="400" src="https://github.com/directus/directus/assets/5363448/0a3ff64a-79f6-467c-9dcb-4a20b8341e4b"></tr></td></table>
- Clean-up of error handler
  - Wrap in async handler, similar to what we're using in other middleware/handlers - makes it easier to read and allows us to easily catch possible errors in the handler itself (unexpected, but would leak filesystem structure info if uncaught)
  - Add comments to reveal some of the magic happening here 🪄 
  - Reusable fallback error
  - Prevent adding undefined extensions data / stack traces to error data (in dev mode)
  - Don't use `toArray` util to transform errors - would split string errors by commas
  - Better typing
- Shared `InternalServerError`
- Add tests

## Potential Risks / Drawbacks

None

## Review Notes / Questions

None

---

Fixes #22420, partially addresses #22416

paescuj committed:
Fix message for unknown errors & additional clean-ups

paescuj committed:
Add changeset

hanneskuettner commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,125 @@
-					stack: error['stack'],
-				};
+			// If available, expose stack trace under error's extensions data
+			if (isObject(error) && error['stack']) {
+				(error['extensions'] ??= {} as any).stack = error['stack'];
```
Normally I'm a fan of short handed assignments, but maybe this is more readable?

```suggestion
			if (isObject(error) && error['stack'] && (error['extensions'] === undefined || isObject(error['extensions']))) {
				error['extensions'] = {
					...error['extensions'],
					stack: error['stack'],
				}
```

hanneskuettner reviewed (commented):
One minor nitpick, otherwise it looks super clean

paescuj committed:
Slightly more readable stack check

Co-authored-by: Hannes Küttner <4376726+hanneskuettner@users.noreply.github.com>

hanneskuettner committed:
Add missing semicolon

hanneskuettner committed:
Remove junk file

paescuj committed:
Only use stack traces from actual `Error`s

br41nslug commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,124 @@
+};
 
-	const errors = toArray<unknown>(err);
+const FALLBACK_ERROR = {
+	status: 500,
+	message: 'An unexpected error occurred.',
+	extensions: {
+		code: 'INTERNAL_SERVER_ERROR',
+	},
+};
```
Could we move this error definition to the `@directus/errors` package? 😄

paescuj replied:
I didn't do so because we don't need a fully fledged `DirectusError` here and `DirectusError` would expect `code` under its root, so we would have to some re-transformation.
But I agree that it would make sense to make it available to clients via `@directus/errors`.
I'll have a look 👍

br41nslug replied:
Yeah exactly for extension use was my initial thinking there 😄

paescuj replied:
Done ✅

paescuj committed:
Add InternalServerError, use in error handler

hanneskuettner committed:
Merge branch 'main' into enhance-api-error-handler

br41nslug commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,118 @@
-			payload.errors.push({
+			errors.push({
 				message: error.message,
 				extensions: {
+					// Expose error code under error's extensions data
 					code: error.code,
 					...(error.extensions ?? {}),
```
```suggestion
				extensions: {
					...(error.extensions ?? {}),
					code: error.code,
```
Inverted the order to ensure the `error.extensions` spreading cannot overwrite the set code.

br41nslug commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,118 @@
+});
+
+function asyncErrorHandler(fn: ErrorRequestHandler) {
+	return (err: any, req: Request, res: Response, next: NextFunction) =>
```
```suggestion
    // Note: keep all 4 parameters here. That's how Express recognizes it's the error handler, even if
    // we don't use next
	return (err: any, req: Request, res: Response, next: NextFunction) =>
```

Lets not forget adding back the original comment here

paescuj replied:
Covered via tests now 👍

br41nslug reviewed (commented):
Looking good, some comments I feel are unnecessary or overkill but they're not hurting anyone so I dont think those matter and would just add noise in the comments 😄

paescuj committed:
Ensure error.code cannot be overwritten

Co-authored-by: Brainslug <tim@brainslug.nl>

paescuj committed:
Catch case where headers have already been sent

paescuj committed:
Enhance asyncErrorHandler signature & handler call

- Omit Promise.resolve as we know we receive a promise in this case
- Use args from ErrorRequestHandler

paescuj committed:
Add tests

paescuj commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,118 @@
+
+function asyncErrorHandler(fn: ErrorRequestHandler) {
+	return (err: any, req: Request, res: Response, next: NextFunction) =>
+		Promise.resolve(fn(err, req, res, next)).catch((error) => {
```
Ah good thought! The function has initially been adapted from https://github.com/directus/directus/blob/7e6980dfdf4a09b06455ac63ae9469dcb382a81a/api/src/utils/async-handler.ts#L3-L4 but since we only use  it here in this file we know that we receive an async function, I've now dropped the `Promise.resolve` wrapping.

paescuj committed:
Merge branch 'main' into enhance-api-error-handler

paescuj committed:
Merge branch 'main' into enhance-api-error-handler

sideshot commented:
Can we get this fixed and pushed through?

rijkvanzanten committed:
Merge branch 'main' into enhance-api-error-handler

br41nslug commented on api/src/middleware/error-handler.ts:
```diff
@@ -1,105 +1,123 @@
+function asyncErrorHandler(
+	fn: (...args: Parameters<ErrorRequestHandler>) => Promise<any>,
+): (...args: Parameters<ErrorRequestHandler>) => Promise<ReturnType<ErrorRequestHandler>> {
+	return (err, req, res, next) =>
```
I dont necessarily think this abstraction is an improvement over my last review but it also doesnt hurt anyone 🤷 Could we however still add back the original comment that this function requires to have 4 parameters to be marked as global error handler 😄

paescuj replied:
What do you mean by abstraction in particular? I just removed the `Promise.resolve` wrapping in d07bed1 since the last review. Any suggestions for improvement?
IMO, the comment is no longer necessary, as it is covered by the new tests and we now (have to) forward all parameters anyway.

br41nslug reviewed (approved):
Not blocking it on adding a comment 😂
